CONTRIBUTING file and code style guidelines (#1822)
fetchai/ledger: docs/cplusplus-style-guide.md

# C++ code style guidelines

We use C++14, but currently consider C++17 too immature for usage.

This documents contains guidelines regarding:

* File structure
* Comments and documentation
* Naming
* Formatting
* Typing
* Namespaces
* Structs
* Classes
* Error handling
* Memory handling
* Resource handling
* Other

## File structure

* Header files: `.hpp`
* Source files: `.cpp`
* Use `#pragma once` in headers.
* Split class definition across a header and source file by default.
* Include header definitions in the same file.

* Order of includes
	1. Related header.
	2. `C` library.
	3. `C++` library.
	4. Other libraries.
	5. Fetch library.

> * Classes in this order

Reviewer: ```suggestion
* Class access specifiers in this order
```

  {
    struct FetchImpl
    {
      /* ... */
    };

    void HelperFunction(FetchImpl &s)
    {
      /* ... */
    }
    };  // namespace details

    class MyFetchClass
    {

    public:
      void DoSomeThing()
      {
        // This improves readability
        using namespace details;

        FetchImpl f;
        HelperFunction(f);
      }
      // ...
    };
  };
};
```

## Structs

Use structs for Plain Old Data (POD) and, in some cases, template meta-programming. Use `classes` for everything else.

``` c++
struct Person
{
  std::string name;
  int         age;
  // ...
};
```

## Classes

* Initialise class members in the following order of preference:
  1. Inline initialisation. This ensures, for instance, that the default constructor does not produce uninitialised memory.
  2. Constructor initialiser list. Member variables must be listed in the order they appear.
  3. Constructor body.

In most cases, a combination of the above will be used. For example:

``` c++
class MyClass
{
public:
  // Inline initialisation
  MyClass() = default;

  // Initialisation list for size, constructor body for pointer
  explicit MyClass(std::size_t n)
    : size_(n)
  {

>     pointer_ = new int[n];

Reviewer: Isn't this assigning an array of ints to a int * here?

Author: Unfortunate given that we discourage the usage of raw pointers. I will change this just to illustrate what the different types of initialisation look like.

Reviewer: 👍

    for (std::size_t i = 0; i < n; ++i)
    {
      pointer_[i] = 0;
    }
  }
  // ...

private:
  std::size_t size_    = 0;
  int *       pointer_ = nullptr;
};
```

* Mark all constructors and assignment operators explicitly, even those which are implicit, to ensure there is no confusion about the intended behaviour of the class.
* Mark unneeded default implementations as deleted. C++ infers whether to create implicit constructors based on whether there are user defined ones. This means that when someone declares a constructor, C++ may remove the default constructor.

``` c++ class Transaction
{
public:
  Transaction()                    = default;
  Transaction(Transaction const &) = default;
  Transaction(Transaction &&)      = deleted;

  ~Transaction() = default;

  Transaction operator(Transaction const &) = default;
  Transaction operator(Transaction &&)      = deleted;
}
```

* Mark a class `final` if it is not meant to be subclassed. The same counts for final implementations of virtual functions.

> ``` c++

Reviewer: Don't think this needs an example (use final)

class Foo
{
public:
  virtual ~Foo()                        = default;
  virtual void Greet()                  = 0;
  virtual void Say(std::string const &) = 0;
};

class Bar : public Foo
{
public:
  void Greet() override final
  {
    // ...
  }

  void Say(std::string const &what) override {
      // ...
  };
};

class Baz final : public Bar
{
public:
  void Say(std::string const &what) override
  {
    // ...
  }
};
```

## Error handling

* Always check error codes and return codes.
* Use exceptions *only* for exceptional circumstances.
* Handle normal errors gracefully.

* Do not rely on exceptions to avoid taking responsibility.
* Do not throw exceptions in destructors as this can crash the program.

## Memory handling

* Use references where the value cannot be null. Use pointers where it can.

> * Make sure no uninitialised memory is accessed.

Reviewer: ```suggestion
* Make sure no uninitialised memory is accessed or created.
```

* Always prefer passing by reference or `const` reference where possible.
* Prefer smart pointers (`std::shared_ptr`, `std::unique_ptr`, etc) over raw pointers.
* Use `make_shared` and `make_unique` to wrap objects in smart pointers.
* Prefer the type-safe `nullptr` over the weakly-typed literal `0` and the `NULL` macro.
* Stick to RAII principle and automatic memory management - destructors are ideally empty.
* When working with array object types, prefer existing structures that already implement memory management:
  * From the standard library, such as `std::vector` and `std::string`.
  * `SharedArray` and `Array` classes from `fetch::memory` namespace.
  * Classes from `fetch::byte_array` namespace, such as `BasicByteArray`.

> * It is safe to transfer ownership from `std::unique_ptr` to `std::shared_ptr`. However, do **NOT** transfer ownership the other way around, as that will result in memory access exceptions; in this case a non-owning `std::weak_ptr` should be used instead of `std::unique_ptr` (see [reference documentation](https://en.cppreference.com/w/cpp/memory/weak_ptr)).

Reviewer: I don't think people will try to do this, I would suggest removing lines 315 - 354

  {
    std::unique_ptr<int> a1 = std::make_unique<int>(5);
    a0                      = a1;  // Transfer of OWNERSHIP from `a1` to `a0`
    // `a1` does NOT control lifecycle of the `int` instance anymore.

    // This will result in throwing an exception since `a1` is now `nullptr`
    std::cout << "a1=" << *a1 << std::endl;

    // The automatically called destructor of `a1` does NOTHING since
    // `a1` does NOT control any object.
  }

  // Automatically called destructor of `a0` destroys the `int` instance
  // allocated and initialised in the nested scope above.
}
```

## Resource handling

* Use a smart pointer to manage resources whose lifecycle needs strict control, such as:
  * Thread synchronisation constructs (e.g. mutex).
  * Filesystem objects.
  * Network connections.
  * Database sessions.

* A destructor must **NEVER** throw an exception. The reason for this is that when the destructor is called automatically because an exception is thrown, if said destructor throws another one, the C++ runtime will call `std::terminate` of the process (as per the [C++ standard definition](http://en.cppreference.com/w/cpp/language/destructor#Exceptions)).

> * Use locking guards provided in C++ `std` namespace, such as `std::unique_lock` and `std::lock_guard`.

Reviewer: We have a macro for locking that appears to be favoured by @ejfitzgerald - FETCH_LOCK - might be worth saying to use this

* Prefer implementations of mutex in the `fetch::mutex` namespace (which conform to a `std` locking API contract) over a plain `std::mutex`. These are `fetch::mutex::DebugMutex` and `fetch::mutex::ProductionMutex`.

## Other

* Don't use `#define`. If you ignore this guidance, choose a `FETCH_STARTING_REALLY_LONG_NAME_THAT_CANNOT_POSSIBLY_CLASH_WITH_ANYTHING_ELSE` and undefine it as soon as you are done with it.
* Use the following keywords when appropriate:
	1. `explicit`.
	2. `override`.
	3. `virtual`.
	4. `final`.
	5. `noexcept`.
* Pre increment: `++i` as it is more performant for non primitive types.
* Lambda captures must be explicit reference or copy.

> * TODOs should refer to the author in the style `TODO(`HUT`) :`.

Reviewer: ```suggestion
* TODOs should refer to the author in the style `TODO(HUT) :`.
```

> # C++ code style guidelines

Reviewer: I have always had it in the back of my mind to rewrite these style guidelines, and now might be the time to do it (?).

To me, they are far too long due to mentioning style choices that are obvious, such as "use const", "use smart pointers"... I think that a document that outlines the ways our style guide differs from others (trailing underscore class variables), or style that developers find contentious (how much to use templates and comments) is more valuable.

Author: I have removed the parts that were closer to a tutorial than style guidelines. The document should be a lot shorter now.

I have preserved those 2 notes in particular, not because I think they are unexpected, but because I think it is good to clarify to newcomers that we are exhaustive about `const` for instance. They are also good introductory points for other things like how do we prefer `const` declarations to look like, particulars about smart pointer handling, etc.

# C++ code style guidelines

We use C++14, but currently consider C++17 too immature for usage.

This documents contains guidelines regarding:

* File structure
* Comments and documentation
* Tests
* Naming
* Formatting
* Typing
* Namespaces
* Structs
* Classes
* Error handling
* Memory handling
* Resource handling
* Other

## File structure

* Header files: `.hpp`
* Source files: `.cpp`
* Use `#pragma once` in headers.
* Split class definition across a header and source file by default.
* Include header definitions in the same file.

* Order of includes
	1. Related header.
	2. `C` library.
	3. `C++` library.
	4. Other libraries.
	5. Fetch library.

* Classes in this order
	1. `public`.
	2. `protected`.
	3. `private`.

## Comments and documentation

* Provide API documentation with your code.
* Include a comment summary at the top of all files.

``` c++

> /* Approximation of the exponential function.

Reviewer 2: I think that this style is a little inconsistent with the current project. I think maybe we find a better example

Author: After the indentation changes etc, does this still look inconsistent?

   * @x is the argument that is used in the exponent 2^(x / M_LN2).
   *
   * We exploit the IEEE-754 standard to efficiently create a linear
   * interpolation between points of the form 2^n where n is an integer.
   *
   * You can find the full details in
   * [1] "A fast, compact approximation of the exponential function",
   * Nicol N. Schraudolph, 1999 Neural computation 11, 853-862.
   *
   * @return the exponential of x.
   */
double Exp(double const &x) 
{

>     // We use the constants explained in [1] and additional make e the

Reviewer 2: Also we are using 2 space indents and that is consistent in this and other examples

    // base rather than using 2.
    static double const a = double(1ull << 20) / M_LN2;
    static double const b = ((1ull << (10)) - 1) * double(1ull << 20); - 60801;
    double in = x * a + b;

    // Move the computed bits into the exponent with spilling into the
    // Mantissa. The spilling creates the linear interpolation, while
    // the exponent creates exact results whenever the argument is an integer.     
    union 
    { 
      double d;
      uint32_t i[2];
    } conv;

    conv.i[1] = static_cast<uint32_t>(in);
    conv.i[0] = 0;

    return conv.d;
}
```

* Comment the non-obvious, not the obvious.

``` c++
// Turns out sorting this list before processing it is faster due to branch 
// prediction, so don't remove this without checking the profile:

> std::sort(users.begin(), users.end(), UserSortPredicate());

Reviewer 2: ```suggestion
std::sort(users.begin(), users.end(), UserSortPredicate);
```

> for (const User& user: users)

Reviewer 2: ```suggestion
for (User const &user: users)
```

{
    if (user.ready)
    {
        // Do something...
        
    }
}
```

* If you do not understand some code, contact the original author. If they have left the company ask the most competent person. After working out what a piece of a non-obvious code is doing, remember to document it.

## Tests

* Write tests and run them
* Do not assume, do assert

## Naming

* Use meaningful names, even in tests.

> * For structs, use all lowercase with underscore to separate words, or camel case, i.e. `first_name` and `firstName` are both valid.

Reviewer 2: This is not consistent. We use `Pascal` case for struct names. Attributes are `snake_case`

Author: Does the same apply to class names? I will document both for the sake of clarity.

Reviewer 2: Yes

> * For classes, private member variables end with an `_`.

Reviewer 2: ```suggestion
* For classes, private and protected member variables end with an `_`.
```

> * `snake_case_` variables with trailing underscore, except in structs where it is `snake_case` without underscore.

Reviewer 2: Should be included in the statement above maybe?

Author: Done. What is the reason for that differentiation?

* For templates, use all uppercase with underscores to separate words. Long and explicit names are preferred to short and obscure names, e.g. `template <typename ARRAY_TYPE>` is preferable to `template <typename A>);`. Short names are acceptable where the type is obvious and reused many times in complex equations.

## Formatting

* Make use of the `apply_style` script provided for formatting.
* Lines should be no longer than 100 characters.

> * Use spaces to indent, not tabs.

Reviewer 2: ```suggestion
* Use 2 spaces to indent, not tabs.
```

* Curly brackets `{...}` should be on their own line.
* There should be a space after conditionals, i.e. `if (test)`
* No spaces inside `(parenthesis)` or `<template>` unless it improves readability.
* No spaces around arrow: `this->fnc();`.
* No trailing whitespace.

> * Enums use `UPPER_CASE`.

Reviewer 2: ```suggestion
* Enums use `UPPER_CASE`. Prefer `enum class` over regular `enum`
```

* Prefer `(u)int(16/32/64)_t` and `std::size_t`.
* Use `static_cast`, `reinterpret_cast`, `dynamic_cast`, etc., rather than C-style casts.
* Avoid `auto` except when the return type is clear.
* Avoid use of RTTI.

* `using` is preferable to `typedef` for type alias definition.
* Make `using` private unless you intend to expose it.
* `namespace {.<template T>` only allowed if `typedef`/`using` indicates the type.

* If a value is `const`, declare it so.
* Use `int const &i` over `const int &i`.

## Namespaces 

* Do not do `using namespace foo` or `namespace baz = ::foo::bar::baz` in `.hpp` files.
* Do not do `using namespace foo` in the top `fetch` namespace, as it pollutes all of it. Instead, do the following:

``` c++
namespace fetch 
{
    namespace details 
    {
        struct FecthImpl { /* ... */ };
      
        void HelperFunction(FecthImpl & s) 
        {
           /* ... */
        }  
    };

    class MyFetchClass {

        public:

            void DoSomeThing() {
              // This improves readability
              using namespace details; 

              FetchImpl f;
              HelperFunction(f);
            }
            // ...
    };

> };

Reviewer 2: The identation should be updated

```

## Structs

Use structs for Plain Old Data (POD) and, in some cases, template meta-programming. Use `classes` for everything else.

``` c++
struct Person 
{
    std::string name;
    int age;
    // ...

> };

Reviewer 2: Alignment and identation

```

## Classes

* Initialise class members in the following order of preference:
	1. Inline initialisation.
	2. Constructor initialiser list.
	3. Constructor body.

* Mark all constructors and assignment operators explicitly, even those which are implicit, to ensure there is no confusion about the intended behaviour of the class.
* Mark unneeded default implementations as deleted. C++ infers whether to create implicit constructors based on whether there are user defined ones. This means that when someone declares a constructor, C++ may remove the default constructor.

``` c++
class Transaction 
{
  public:
      Transaction() = default;
      Transaction(Transaction const&) = default;
      Transaction(Transaction &&) = deleted;

      ~Transaction() = default;

      Transaction operator(Transaction const&) = default;
      Transaction operator(Transaction &&) = deleted;

> }

Reviewer 2: Identation - specifically we don't align the `public` /  `private` keywords either

```

* Mark a class `final` if it is not meant to be subclassed. The same counts for final implementations of virtual functions.

``` c++
class Foo 
{
  public:

>       virtual ~Foo() { }

Reviewer 2: ```suggestion
      virtual ~Foo() = default;
```

      virtual void Greet() = 0;
      virtual void Say(std::string const&) = 0;   
};

class Bar : public Foo 
{
  public:
      void Greet() override final 
      {
        // ...
      }
      
      void Say(std::string const& what) override 
      {
        // .. 
      };     
};

class Baz final : public Bar 
{
  public:
      void Say(std::string const& what) override 
      {
        // ...
      } 

> }; 

Reviewer 2: Identation

```

> C++14 allows for inline initialisation of member variables which should be preferred at all times as uninitialised memory can have secondary effects making it almost impossible to debug a problem. 

Reviewer 2: ```suggestion
Inline initialisation of member variables is preferred at all times as uninitialised memory can have secondary effects making it almost impossible to debug a problem. 
```

When it is not possible to use inline initialisation, member variables should be initialised using the constructor initialisation list where they must be listed in the order they appear. 

Last resort is initialisation inside the constructors body. In most cases, a combination of the above will be used. For example:

``` c++
class MyClass 
{
  public:
      MyClass() = default;
      

>       MyClass(std::size_t n)

Reviewer 2: ```suggestion
      explicit MyClass(std::size_t n)
```

        : size_(n) 
      {
        pointer_ = new int[n];
        for(std::size_t i=0; i< n; ++i)
        {
          pointer_[i] = 0;    
        }
      }
      // ...
  private:
      std::size_t size_ = 0;

>       int* pointer_ = nullptr;

Reviewer 2: Alignment

Example of using smart pointer with **shared** ownership:

``` c++
#include <memory>
#include <iostream>

//* Parent Scope block
{
  std::shared_ptr<int> a0;
  
  //* Nested scope block
  {
    std::shared_ptr<int> a1 = std::make_shared<int>(5); //* refcount=1
    a0 = a1; //* The `a1` instance become shared at this point (refcount=2)
    
    //* BOTH instances are valid, controlling the same `int` instance:
    std::cout << "a0 = " << *a0 << ", a1 = " << *a1 << std::endl;
    
    throw std::exception();
    
    //* The exception will cause stack unwinding, during which 
    //* is the destructor of `a1` instance called automatically, 
    //* what decrements refcount of shared object (to 1).
    //* At this point, the `a0` shared object is controlling lifecycle
    //* of the `int` instance created above in this scope.
  }
  
  //* We are still in stack unwinding process here (due to earlier exception),
  //* during which automatically called destructor of `a0` instance  decrements
  //* refcount of shared object (to 0), what finally results to destruction of

>   //* the `int` instance allocated & initialised in the nested scope above.

Reviewer 2: This `// *` syntax probably should be replaced with `//` to be more consistent

}
```

Example of using smart pointer with **exclusive** ownership:
``` c++
#include <memory>

//* Parent Scope block
{
  std::unique_ptr<int> a0;
  

>   //* Nested scope block

Reviewer 2: And again here - also not sure if we want to include this section as it basically an copy and paste from the one above
